debezium-core/src/main/java/io/debezium/pipeline/ChangeEventSourceCoordinator.java

```diff
@@ -256,10 +322,73 @@ public ErrorHandler getErrorHandler() {
 
     public class ChangeEventSourceContextImpl implements ChangeEventSourceContext {
 
+        private final Lock lock = new ReentrantLock();
+        private final Condition snapshotFinished = lock.newCondition();
```

Comment:
Is it guranteed that two threads will not end up in a deadlock waiting on the conditions?

Reply:
The Snapshot thread and the streaming thread are never waiting on the same condition. Only one of two is in the wait.